src/garage/_dtypes.py

```diff
@@ -167,3 +166,128 @@ def __new__(cls, env_spec, observations, actions, rewards, terminals,
         return super().__new__(TrajectoryBatch, env_spec, observations,
                                actions, rewards, terminals, env_infos,
                                agent_infos, lengths)
+
+
+class Sample(
```

Comment:
i don't think the name of this type should be `Sample` because:

1. The name is already overloaded. For instance, when you say sample, do you mean the verb "to sample"? Or do you mean sampling from a distribution? or sampling from an environment? It's important that the codebase be easily grep-able. Data types should clearly be nouns, not verbs.
1. We don't use the term sample already anyway, no need to add ambiguity
1. We already use the term `TotalEnvSteps` to count these in algorithms. Whatever we choose here should be consistent with that name.

Whatever we use should also be consistent with the idea of a trajectory -- in principle you should be able to transform a bunch of time-ordered Samples into a TrajectoryBatch.

Candidate names:
* Step -- already used by gym
* TimeStep (used by tf_agents)
* EnvStep
* EnvSample

Of these I think that `TimeStep` and `EnvStep` make the most sense. `EnvStep` is kind of redundant (where else do steps come from?) so I suggest you name this `TimeStep` and also refactor usages of `TotalEnvSteps` to read `TotalTimeSteps`.